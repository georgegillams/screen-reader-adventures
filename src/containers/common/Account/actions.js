import { createRoutine } from 'redux-saga-routines';

import { LOGOUT, REQUEST_VERIFICATION_EMAIL } from './constants';

export const logout = createRoutine(LOGOUT);
export const requestVerificationEmail = createRoutine(REQUEST_VERIFICATION_EMAIL);
