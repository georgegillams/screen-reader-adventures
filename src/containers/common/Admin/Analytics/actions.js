import { createRoutine } from 'redux-saga-routines';

import { LOAD } from './constants';

export const load = createRoutine(LOAD);
