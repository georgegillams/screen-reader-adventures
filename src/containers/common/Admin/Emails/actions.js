import { createRoutine } from 'redux-saga-routines';

import { LOAD, RESEND } from './constants';

export const load = createRoutine(LOAD);
export const resend = createRoutine(RESEND);
