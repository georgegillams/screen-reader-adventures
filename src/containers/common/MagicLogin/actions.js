import { createRoutine } from 'redux-saga-routines';

import { LOGIN } from './constants';

export const login = createRoutine(LOGIN);
