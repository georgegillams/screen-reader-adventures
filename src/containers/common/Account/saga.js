import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { logout, requestVerificationEmail } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogout() {
  const requestURL = apiStructure.logout.fullPath;

  try {
    yield put(logout.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(logout.failure(result));
    } else {
      yield put(logout.success(result));
      yield put(setUser.trigger(null));
    }
  } catch (err) {
    yield put(logout.failure(err));
  } finally {
    yield put(logout.fulfill());
  }
}

export function* doRequestVerificationEmail() {
  const requestURL = apiStructure.requestVerificationEmail.fullPath;

  try {
    yield put(requestVerificationEmail.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
    });

    if (result.error) {
      yield put(requestVerificationEmail.failure(result));
    } else {
      yield put(requestVerificationEmail.success(result));
    }
  } catch (err) {
    yield put(requestVerificationEmail.failure(err));
  } finally {
    yield put(requestVerificationEmail.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(logout.TRIGGER, doLogout);
  yield takeLatest(requestVerificationEmail.TRIGGER, doRequestVerificationEmail);
}
