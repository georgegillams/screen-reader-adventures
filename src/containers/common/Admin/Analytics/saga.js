import { takeLatest, put, call } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load } from './actions';

export function* doLoad() {
  const requestURL = apiStructure.loadAnalyticsSummary.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.analytics));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
}
