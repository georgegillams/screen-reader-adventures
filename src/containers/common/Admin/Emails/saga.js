import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { load, resend } from './actions';
import { selectState } from './selectors';

export function* doLoad() {
  const requestURL = apiStructure.loadEmails.fullPath;

  try {
    yield put(load.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(load.failure(result));
    } else {
      yield put(load.success(result.emails));
    }
  } catch (err) {
    yield put(load.failure(err));
  } finally {
    yield put(load.fulfill());
  }
}

export function* doResend() {
  const currentState = yield select(selectState());
  const { emailToResend } = currentState;
  const requestURL = apiStructure.resendEmail.fullPath;

  try {
    yield put(resend.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(emailToResend),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(resend.failure(result));
    } else {
      yield put(resend.success(result.emails));
      yield put(load.trigger());
    }
  } catch (err) {
    yield put(resend.failure(err));
  } finally {
    yield put(resend.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(load.TRIGGER, doLoad);
  yield takeLatest(resend.TRIGGER, doResend);
}
