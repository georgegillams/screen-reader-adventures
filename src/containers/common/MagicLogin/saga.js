import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import { login } from './actions';
import { setUser } from 'containers/common/Authenticator/actions';

export function* doLogin() {
  const currentState = yield select(selectState());
  const { token } = currentState;
  const requestURL = apiStructure.loginWithMagicLink.fullPath;

  try {
    yield put(login.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ magicLinkKey: token }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(login.failure(result));
    } else {
      yield put(login.success(result));
      yield put(setUser.trigger(result));
    }
  } catch (err) {
    yield put(login.failure(err));
  } finally {
    yield put(login.fulfill());
  }
}
export default function* saga() {
  yield takeLatest(login.TRIGGER, doLogin);
}
