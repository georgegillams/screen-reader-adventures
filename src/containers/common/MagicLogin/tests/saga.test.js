import { put, takeLatest } from 'redux-saga/effects';

import { login } from '../actions';

import saga, { doLogin } from '../saga';

describe('MagicLogin saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should login on login TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(login.TRIGGER, doLogin));
  });

  describe('account actions', () => {
    let logoutGenerator;

    const response = {
      success: 'you are logged out',
      status: 200,
    };

    beforeEach(() => {
      logoutGenerator = doLogin();

      const selectDescriptor = logoutGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call login.success on successful API call', () => {
      logoutGenerator.next({ token: 'asdfghjkl' });
      logoutGenerator.next();
      const putSuccess = logoutGenerator.next(response).value;
      logoutGenerator.next();

      expect(putSuccess).toEqual(put(login.success(response)));
    });

    it('Should call login.failure if API returns error', () => {
      logoutGenerator.next({ token: 'asdfghjkl' });
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      logoutGenerator.next();
      const putSuccess = logoutGenerator.next(response).value;
      logoutGenerator.next();

      expect(putSuccess).toEqual(put(login.failure(response)));
    });

    it('Should call login.failure if an exception occurs', () => {
      logoutGenerator.next({ token: 'asdfghjkl' });
      const response = new Error('Some error');
      const putFailure = logoutGenerator.throw(response).value;

      expect(putFailure).toEqual(put(login.failure(response)));
    });
  });
});
