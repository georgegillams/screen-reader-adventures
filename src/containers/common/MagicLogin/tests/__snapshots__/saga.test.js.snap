// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicLogin saga account actions Should call login.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`MagicLogin saga account actions Should call login.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`MagicLogin saga account actions Should call login.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
